docs/0002-N-tuple-tree.md

```diff
@@ -0,0 +1,180 @@
+
+## Overview
+
+This extension provides a general mechanism for describing a set of algorithms for mapping numerical identifers onto a tree structure for constructing an OCFL Storage Hierarchy in a way that maintains performance. The aim is to balance the problem of having too many files or subdirectories in any one directory with having an overly deeply nested set of subdirectories. As both identifier formats and specific filesystem instances may vary there is not a single optimal approach.       
+
+## Parameters
+
+* name: identifier length
+  * description: The number of meaningful characters in the base identifier 
+  * type: integer
+  * range: 0,4096
+  * default:
+* name: case mapping
+  * description: Indicates how case in the source identifier should be handled when mapping to a path
+  * type: enumerated
+  * range: ToUpper,ToLower,Literal
+  * default:
+* name: invert mapping
+  * description: Indicate if mapping should begin at the rightmost (least significant) character of the stripped identifier
+  * type: enumerated
+  * range: FALSE,TRUE
+  * default: FALSE
+* name: tuple size
+  * description: Indicates the size of the chunks (in characters) that the identifier is split into during mapping
+  * type: integer
+  * range: 1,32
+  * default: 2
+* name: number of tuples
+  * description: Indicates how many chunks are used for path generation
+  * type: integer
+  * range: 0,32
+  * default:
+* short object root 
+  * description: Indicates how the OCFL object root directory name should be generated from the identifier
+  * type: enumerated
+  * range: FALSE,TRUE
+  * default: FALSE
+
+## Detailed explanation
+
+The approach described here is a generalisation of the [PairTree](https://tools.ietf.org/html/draft-kunze-pairtree-01) algorithm designed to be more flexible as file storage technologies have developed. Conventional filesystems are generally better able to handle large numbers of files in a directory and object stores tend to favour much flatter storage hierarchies. In short, the approach is to derive a unique file path for an OCFL object from its unique identifier in a programmatic and repeatable manner. 
+
+### identifier length
+
+This extension assumes that object unique identifiers are all the same length and that all the characters of the identifer are reasonably well distributed (e.g. for a hexadecimal-based identifier, each character can be any value from 0-f). This may mean that it is prudent to adjust the format of the identifier before it can be safely used to generate a path. For example, a [UUID](https://tools.ietf.org/html/rfc4122) is typically written in the form "uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6" where the "uuid:" premable and the hyphens are clearly non-unique elements. The "stripped" version used for path generation would be "f81d4fae7dec11d0a76500a0c91e6bf6". The **identifier length** parameter indicates the length of the stripped version of the identifier, in this example, 32.   
```

Comment:
I think uniform id length is an unnecessary restriction. Why should all identifiers be the same length? What if I change my id scheme?

I understand that this is a key simplification of pairtree to avoid addressing annoying edge cases (eg objects`ab` and `abc`), but the result is a spec that is narrowly applicable to well formed identifiers only.

Is this parameter intended to be used as an id validation? That is, reject ids that aren't `n` characters long? If so, that behavior should be described.

Reply:
Correct, fixed length is intended. Any spec that doesn't allow consistent length will have to deal with the edge cases and that is a lot longer and should be a different extension. Like OCFL this is a human readable specification for a static layout.

Reply:
In that case, maybe this extension should be renamed something like "Fixed Length Identifier N-tuple Trees for OCFL Storage Hierarchies". Personally, I don't think this extension is particular useful or forward thinking due to the inability to change the length of your identifier scheme without creating a new OCFL repository.

Reply:
Variable length identifiers potentially result in objects in hierarchy intermediate directories which is not OCFL compliant.

Reply:
OCFL also requires one layout pattern for it's storage hierarchies. An extension that defines a variable pattern is not really in the spirit either.

Reply:
I'm confused. So, you're saying that the pairtree spec is not an OCFL complaint way to map object roots?

That a layout like this is unacceptable:

```
   ab/
   |
   \--- cd/
        |
        |--- foo/
        |    |   0=ocfl_object_1.0
        |    |    ...
        |    |
        |    \--- gh/
        |
        \--- e/
             |
             \--- bar/
                  |  0=ocfl_object_1.0
                  |   ...
```

Lifted from [here](https://tools.ietf.org/html/draft-kunze-pairtree-01#section-2).

Reply:
The spec says... "OCFL Object Roots MUST be stored either as the terminal resource at the end of a directory storage hierarchy or as direct children of a containing OCFL Storage Root."

Reply:
I don't see where the contradiction is.

I don't like pairtree, but let's continue to use it as an example. If you have variable length object identifiers and you directly map these identifiers to object root directories using the pairtree spec, then the result will be objects at variable depths. But, every object will be encapsulated, every terminal node of the storage hierarchy will be an OCFL object, and no OCFL object will contain another OCFL object.

Reply:
..and some objects will not be either terminal resources or direct children of the storage root.

Reply:
What is your definition of a "terminal resource"? Is a directory that encapsulates an OCFL object in its entirety and does not contain any other OCFL objects, not a "terminal resource"?

Reply:
Right - yes that's fine. Just not covered by this one.

Reply:
Renamed to indicate fixed length identifiers in the new PR. Handling the wrinkles of variable length identifiers can be left for another day and extension!